File: backend/app/api/radar.py
"""Роутер /api/radar/* — раздел разведки по конкурентам (docs/specs/08-radar.md).

Тонкий роутер: валидация запроса + делегирование в app/radar/{scrape_service,analyze_service,
analytics,report,repo}.py. Разбор (`/jobs/*`) вызывает функции агента B из app.radar.pipeline —
импорт лениво, внутри обработчиков: тесты подменяют app.radar.pipeline в sys.modules.

Сверено с реализацией B (app/radar/pipeline.py, 2026-09-24): claim_job и claim_stale_job —
обычные синхронные функции (захват задания — один условный update, ждать нечего), а
process_radar_job и refresh_expired_links — корутины (внутри скачивание видео и вызов Gemini).
"""
from __future__ import annotations

from typing import Any, Optional

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import Response

from app.models.radar_schemas import (
    AnalyzeCancelRequest,
    AnalyzeRequest,
    RefreshFollowersRequest,
    ScrapeRequest,
)
from app.radar import analytics, analyze_service, repo, report, scrape_service
from app.radar import settings as radar_settings

router = APIRouter(prefix="/radar", tags=["radar"])


# ── Конфигурация фронта ──────────────────────────────────────────────────

@router.get("/config")
async def get_config() -> dict[str, Any]:
    return {"whisper_available": bool(radar_settings.OPENAI_API_KEY)}


# ── Конкуренты ─────────────────────────────────────────────────────────────

@router.get("/competitors")
async def get_competitors() -> list[dict[str, Any]]:
    return repo.list_competitors()


@router.post("/competitors/refresh-followers")
async def refresh_followers(req: RefreshFollowersRequest) -> dict[str, Any]:
    try:
        run_id = await scrape_service.start_followers_scrape(req.usernames)
    except scrape_service.RadarValidationError as exc:
        raise HTTPException(400, str(exc)) from exc
    except scrape_service.RadarHourlyLimitError as exc:
        raise HTTPException(429, str(exc)) from exc
    except scrape_service.RadarApifyError as exc:
        raise HTTPException(502, f"Apify не ответил: {exc}") from exc
    return {"run_id": run_id}


# ── Сбор ───────────────────────────────────────────────────────────────────

@router.get("/scrape/cost-estimate")
async def scrape_cost_estimate(
    usernames: str = Query(""),
    period_months: int = Query(3),
) -> dict[str, Any]:
    names = scrape_service.normalize_usernames(usernames.split(","))
    return scrape_service.estimate_cost(names, period_months)


@router.post("/scrape")
async def start_scrape(req: ScrapeRequest) -> dict[str, Any]:
    try:
        run_id = await scrape_service.start_reel_scrape(req.usernames, req.period_months)
    except scrape_service.RadarValidationError as exc:
        raise HTTPException(400, str(exc)) from exc
    except scrape_service.RadarHourlyLimitError as exc:
        raise HTTPException(429, str(exc)) from exc
    except scrape_service.RadarApifyError as exc:
        raise HTTPException(502, f"Apify не ответил: {exc}") from exc
    return {"run_id": run_id}


@router.get("/scrape/{run_id}")
async def scrape_status(run_id: int) -> dict[str, Any]:
    run = await scrape_service.poll_scrape_run(run_id)
    if run is None:
        raise HTTPException(404, "Прогон не найден")
    return {
        "id": run["id"],
        "kind": run["kind"],
        "status": run["status"],
        "results_count": run.get("results_count"),
        "cost_estimate_usd": run.get("cost_estimate_usd"),
        "error": run.get("error"),
        "created_at": run.get("created_at"),
    }


# ── Аналитика ──────────────────────────────────────────────────────────────

@router.get("/summary")
async def summary(usernames: Optional[str] = Query(None)) -> dict[str, Any]:
    return analytics.get_summary(usernames)


@router.get("/reels")
async def get_reels(
    usernames: Optional[str] = Query(None),
    sort: str = Query("views"),
    order: str = Query("desc"),
    min_views: Optional[int] = Query(None),
    max_views: Optional[int] = Query(None),
    limit: int = Query(200, le=1000),
) -> list[dict[str, Any]]:
    return analytics.get_reels(
        usernames=usernames, sort=sort, order=order,
        min_views=min_views, max_views=max_views, limit=limit,
    )


# ── Разбор ─────────────────────────────────────────────────────────────────

@router.post("/analyze")
async def start_analyze(req: AnalyzeRequest) -> dict[str, Any]:
    if not req.items:
        raise HTTPException(400, "items required")

    try:
        return await analyze_service.enqueue_analysis(req.items, req.force, req.transcriber)
    except analyze_service.RadarReelNotFoundError as exc:
        raise HTTPException(404, str(exc)) from exc
    except analyze_service.RadarAnalysisDailyLimitError as exc:
        raise HTTPException(429, str(exc)) from exc
    except analyze_service.RadarWhisperUnavailableError as exc:
        raise HTTPException(400, str(exc)) from exc


@router.post("/jobs/{reel_id}/run")
async def run_job(reel_id: str) -> dict[str, Any]:
    from app.radar.pipeline import claim_job, process_radar_job

    job = claim_job(reel_id)
    if job is None:
        # Уже захвачено другим вызовом (или не queued) — отдаём текущий статус, не разбираем повторно.
        analysis = repo.get_analysis(reel_id)
        status = analysis["status"] if analysis else "not_started"
        return {"reel_id": reel_id, "status": status}

    status = await process_radar_job(job)
    return {"reel_id": reel_id, "status": status}


@router.post("/jobs/tick")
async def jobs_tick() -> dict[str, Any]:
    from app.radar.pipeline import claim_stale_job, process_radar_job

    job = claim_stale_job()
    if job is None:
        return {"picked": None, "status": None}

    status = await process_radar_job(job)
    return {"picked": job["reel_id"], "status": status}


@router.post("/analyze/cancel")
async def cancel_analyze(req: AnalyzeCancelRequest) -> dict[str, Any]:
    if not req.reel_ids:
        raise HTTPException(400, "reel_ids required")
    return {"cancelled": repo.cancel_jobs(req.reel_ids)}


@router.get("/analyze/status")
async def analyze_status(reel_ids: str = Query(...)) -> dict[str, Any]:
    ids = [r.strip() for r in reel_ids.split(",") if r.strip()]
    if not ids:
        raise HTTPException(400, "reel_ids required")

    rows = repo.get_analyses_by_ids(ids)
    result: dict[str, Any] = {}
    for reel_id in ids:
        row = rows.get(reel_id)
        if row:
            result[reel_id] = {
                "reel_id": reel_id,
                "mode": row.get("mode"),
                "status": row.get("status"),
                "error": row.get("error"),
                "updated_at": row.get("updated_at"),
            }
        else:
            result[reel_id] = {
                "reel_id": reel_id, "mode": None,
                "status": "not_started", "error": None, "updated_at": None,
            }
    return result


@router.get("/reels/{reel_id}/analysis")
async def reel_analysis(reel_id: str) -> dict[str, Any]:
    reel = repo.get_reel(reel_id)
    if reel is None:
        raise HTTPException(404, "Рилс не найден")
    return {"reel": reel, "analysis": repo.get_analysis(reel_id) or {}}


@router.get("/reels/{reel_id}/report.md")
async def reel_report_md(reel_id: str) -> Response:
    reel = repo.get_reel(reel_id)
    if reel is None:
        raise HTTPException(404, "Рилс не найден")
    md = report.build_report_md(reel, repo.get_analysis(reel_id))
    return Response(
        content=md,
        media_type="text/markdown; charset=utf-8",
        headers={"Content-Disposition": f'attachment; filename="report_{reel_id}.md"'},
    )


@router.get("/export.md")
async def export_md(reel_ids: str = Query(...)) -> Response:
    ids = [r.strip() for r in reel_ids.split(",") if r.strip()]
    if not ids:
        raise HTTPException(400, "reel_ids required")

    reels = repo.get_reels_by_ids(ids)
    analyses = repo.get_analyses_by_ids(ids)
    parts = [report.build_report_md(reels[rid], analyses.get(rid)) for rid in ids if rid in reels]

    return Response(
        content="\n\n---\n\n".join(parts),
        media_type="text/markdown; charset=utf-8",
        headers={"Content-Disposition": 'attachment; filename="reels_report.md"'},
    )

File: frontend/src/api/radar.js
// Клиент API раздела «Радар». База — как в client.js: на проде фронт и бэк на
// одном домене (VITE_API_URL не задаётся, запросы идут на относительный
// /api/radar), локально бэкенд поднят отдельно и VITE_API_URL указывается явно
// (см. CLAUDE.md проекта). Разбор ошибок — тоже как в client.js: detail из тела
// ответа, иначе человеческий текст по коду статуса.
const BASE = (import.meta.env.VITE_API_URL || '') + '/api/radar'

const STATUS_MESSAGES = {
  400: 'Некорректный запрос (400)',
  401: 'Требуется авторизация (401)',
  403: 'Доступ запрещён (403)',
  404: 'Не найдено (404)',
  408: 'Истекло время ожидания (408)',
  409: 'Конфликт данных (409)',
  422: 'Некорректные данные (422)',
  429: 'Слишком много запросов, попробуйте позже (429)',
  500: 'Ошибка сервера (500)',
  502: 'Сервер недоступен (502)',
  503: 'Сервис временно недоступен (503)',
  504: 'Сервер не отвечает (504)',
}

function statusMessage(status) {
  return STATUS_MESSAGES[status] || `Ошибка запроса (${status})`
}

async function req(path, opts = {}) {
  let r
  try {
    r = await fetch(BASE + path, opts)
  } catch (e) {
    if (e?.name === 'AbortError') throw e
    throw new Error('Нет соединения с сервером')
  }
  if (!r.ok) {
    let message = statusMessage(r.status)
    try {
      const txt = await r.text()
      if (txt) {
        try {
          const data = JSON.parse(txt)
          if (data && typeof data.detail === 'string' && data.detail.trim()) {
            message = data.detail
          } else if (data && data.detail != null) {
            message = JSON.stringify(data.detail)
          }
        } catch (_) {
          // тело не JSON — оставляем сообщение по статусу
        }
      }
    } catch (_) {
      // не удалось прочитать тело — оставляем сообщение по статусу
    }
    throw new Error(message)
  }
  const ct = r.headers.get('content-type') || ''
  return ct.includes('json') ? r.json() : r.blob()
}

const postJson = body => ({
  method: 'POST',
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify(body),
})

// GET /competitors — список ранее собранных конкурентов с подписчиками
export const getCompetitors = () => req('/competitors')

// GET /scrape/cost-estimate — оценка стоимости сбора на Apify до запуска
export const getCostEstimate = (usernames, periodMonths) =>
  req(`/scrape/cost-estimate?usernames=${encodeURIComponent(usernames.join(','))}&period_months=${periodMonths}`)

// POST /scrape — запуск платного сбора метаданных рилсов
export const startScrape = (usernames, periodMonths) =>
  req('/scrape', postJson({ usernames, period_months: periodMonths }))

// POST /competitors/refresh-followers — обновить подписчиков (без usernames — для всех)
export const refreshFollowers = usernames =>
  req('/competitors/refresh-followers', postJson(usernames?.length ? { usernames } : {}))

// GET /scrape/{run_id} — статус прогона сбора; побочный эффект на бэке — приём датасета
export const getScrapeStatus = runId => req(`/scrape/${runId}`)

// GET /summary — сводка и корзины по просмотрам для набора аккаунтов
export const getSummary = usernames =>
  req(`/summary${usernames?.length ? `?usernames=${encodeURIComponent(usernames.join(','))}` : ''}`)

// GET /reels — лента рилсов с полями analysis_status/analysis_mode
export const getReels = (usernames, sort = 'views', opts = {}) => {
  const qs = new URLSearchParams()
  if (usernames?.length) qs.set('usernames', usernames.join(','))
  qs.set('sort', sort)
  qs.set('order', opts.order || 'desc')
  qs.set('limit', opts.limit ?? 200)
  if (opts.minViews != null) qs.set('min_views', opts.minViews)
  if (opts.maxViews != null) qs.set('max_views', opts.maxViews)
  return req(`/reels?${qs}`)
}

// POST /analyze — постановка рилсов в очередь на разбор (t|v|tv)
export const startAnalyze = (items, force = false) =>
  req('/analyze', postJson(force ? { items, force } : { items }))

// POST /jobs/{reel_id}/run — синхронный прогон одного задания (до ~90 с);
// таймаут fetch намеренно не ставим короче 120 с — тут его просто нет.
export const runJob = reelId => req(`/jobs/${reelId}/run`, { method: 'POST' })

// POST /jobs/tick — страховка для зависших заданий; вызывается pg_cron,
// фронтом не используется, но входит в контракт эндпоинтов ТЗ.
export const tickJob = () => req('/jobs/tick', { method: 'POST' })

// POST /analyze/cancel — отмена ещё не завершённых заданий разбора
export const cancelAnalyze = reelIds => req('/analyze/cancel', postJson({ reel_ids: reelIds }))

// GET /analyze/status — статусы пачки заданий для поллинга
export const getAnalyzeStatus = reelIds =>
  req(`/analyze/status?reel_ids=${encodeURIComponent(reelIds.join(','))}`)

// GET /reels/{id}/analysis — рилс + результат разбора для отчёта
export const getReelAnalysis = id => req(`/reels/${id}/analysis`)

// GET /reels/{id}/report.md — прямая ссылка на скачивание (используется как href)
export const reportUrl = id => `${BASE}/reels/${id}/report.md`

// GET /export.md?reel_ids= — сводный отчёт по пачке рилсов (используется как href)
export const exportUrl = reelIds => `${BASE}/export.md?reel_ids=${encodeURIComponent(reelIds.join(','))}`
